content/blog/2019-10-23-tail-call-elimination.md

```diff
@@ -0,0 +1,239 @@
+```
+then.6:
+  ...
+  call foo;
+  v17: int = retval ;
+  result: int = id v17;
+  jmp endif.6;
+else.6:
+  ...
+  call foo;
+  v21: int = retval ;
+  result: int = id v21;
+endif.6:
+  v22: int = id result;
+  ret v22;
+```
+After copy propagation, it looks like this:
+```
+then.6:
+  ...
+  call foo;
+  result: int = retval;
+  jmp endif.6;
+else.6:
+  ...
+  call foo;
+  result: int = retval;
+endif.6:
+  ret result;
+```
+
+Then we can analyze the CFG backwards to see that indeed we can replace
+the calls to a `jmp`.
+```
+then.6:
+  ...
+  jmp foo;
+  jmp endif.6;
+else.6:
+  ...
+  jmp foo;
+endif.6:
+  ret result;
+```
+Note that the extra instructions can simply be removed by a DCE pass, so we don't
+worry about that.
+
+*Unfortunately, I couldn't get this to work properly because my copy propagation
+pass had bugs.*
+
+## Evaluation
+To evaluate that the tail call elimination is working and actually gives us an
+improvement, we benchmark some recursive functions that use tail recursion, and
+show the difference in execution time and memory usage between an optimized and
+unoptimized Bril program.
+
+The table entries show how much change was observed, as a percentage, by doing
+tail call elimination (TCE). For example, an entry of -10% means the optimized program
+used 10% less memory/time than the unoptimized program. An `X` means that the output of the program was too
+big to handle. `n` is the argument passed to the recursive function.
+`loop` is a Bril program that simply loops `n` times using recursion. `factorial`
+is a tail recursive implementation of factorials. `mutual_rec` is a program that
+checks whether a program is even or odd in a mutually recursive way. The code for
+these can be found [here](https://github.com/sampsyo/bril/pull/37).
+
+**Percentage Change in Memory Usage Using TCE**
+
+|            |   n = 1  |  n = 100 | n = 10000 | n = 100000 |
+|:----------:|:--------:|:--------:|:---------:|:----------:|
+|    loop    |   +0.1%  |   -2.7%  |   -31.1%  |   -79.5%   |
+|  factorial |   +0.2%  |   -2.5%  |   -79.1%  |     X      |
+| mutual_rec |   +0.2%  |   +13.8% |   -31.2%  |   -78.4%   |
+
+**Percentage Change in Execution Time Using TCE**
+
+|            |   n = 1  |  n = 100 | n = 10000 | n = 100000 |
+|:----------:|:--------:|:--------:|:---------:|:----------:|
+|    loop    |  +2.1%   |   +2.2%  |   -10%    |   -29.8%   |
+|  factorial |  +1.1%   |   +5.5%  |   -1.6%   |      X     |
+| mutual_rec |  +1.1%   |   -1%    |   +5.7%   |   +5.07%   |
+
+To get the execution time and peak memory usage, I use `/usr/bin/time -l` (which prints the contents of rusage).
+To make sure the measurements are meaningful, I chose a maximum `n` value so
+that the tests took a few seconds. Here we can clearly see that with large values
+of `n`, the programs with TCE use considerably less memory. However, it is unclear
+whether there is a benefit to the execution time of the program since the values
+vary quite a bit.
```

Comment:
If numbers vary a lot, it might be a good idea to report the variance or standard deviation or whatever to give a more complete picture of the distribution. (No need to go back and do this for this post, of course.)

Reply:
Good point. It definitely would have been good to include those things.